File: cpp/3343.cpp
#include <iostream>
#include <string>

using namespace std;

int main()
{
    int n, x, t;
    cin >> n >> x;
    string titans;
    getchar();
    getline(cin, titans);
    int p, m, g;
    cin >> p >> m >> g;
    int walls[n];
    for (t = 0; t < n; t++) walls[t] = x;
    int i = 0, j = 0, k = 0, sum = 0;

    for (t = 0; t < n; t++) {
        switch (titans[t])
        {
            case 'P':
                while (walls[i] < p) i++;
                walls[i] -= p;
                break;

            case 'M':
                while (walls[j] < m) j++;
                walls[j] -= m;
                break;

            case 'G':
                while (walls[k] < g) k++;
                walls[k] -= g;
                break;
        }
    }

    for (t = 0; t < n; t++) {
        if (walls[t] < x) {
            sum++;
        }
    }

    cout << sum << '\n';
    return 0;
}

File: py/3343.py
def aot():
    n, x = [int(i) for i in input().split()]
    titans = input()
    p, m, g = [int(i) for i in input().split()]
    walls = [x for _ in range(n)]
    i, j, k = 0, 0, 0

    for t in titans:
        if t == 'P':
            while walls[i] < p: i += 1
            walls[i] -= p

        elif t == 'M':
            while walls[j] < m: j += 1
            walls[j] -= m

        else:
            while walls[k] < g: k += 1
            walls[k] -= g

    print(sum([1 for w in walls if w < x]))


aot()